lxd/device/gpu.go

```diff
@@ -163,3 +185,464 @@ func gpuSelected(device config.Device, gpu api.ResourcesGPUCard) bool {
+	revert := revert.New()
+	defer revert.Fail()
+
+	for deviceName, conf := range devicesConfig {
+		switch conf["type"] {
+		case "unix-char":
+			charDev := &unixCommon{
+				deviceCommon: deviceCommon{
+					state:  s,
+					inst:   inst,
+					name:   deviceName,
+					config: conf,
+				},
+			}
+
+			err = chardevHandlerFunc(charDev, rootRunConf)
+			if err != nil {
+				return err
+			}
+
+		case "disk":
+			disk := &disk{
+				deviceCommon: deviceCommon{
+					state:  s,
+					inst:   inst,
+					name:   deviceName,
+					config: conf,
+				},
+			}
+
+			err = diskdevHandlerFunc(disk, rootRunConf)
+			if err != nil {
+				return err
+			}
+
+		default:
+			return fmt.Errorf("CDI device of wrong type: %s", conf["type"])
+		}
+	}
+
+	revert.Success()
+	return nil
+}
+
+// constructCDIMountList builds a list of disk devices to be created from a CDI spec.
+func constructCDIMountList(nativeDevices map[string]map[string]string, cdiID cdi.ID, mounts []*specs.Mount, instPrivileged bool) ([]cdi.SymlinkEntry, error) {
+	if len(mounts) == 0 {
+		return nil, fmt.Errorf("CDI mounts are empty")
+	}
+
+	indirectSymlinks := make([]cdi.SymlinkEntry, 0)
+	var splittedDevName []string
+	var chosenOpts []string
+	var deviceName string
+	var err error
+
+	shift := "false"
+	if instPrivileged {
+		shift = "true"
+	}
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	for _, mount := range mounts {
+		if mount.HostPath == "" || mount.ContainerPath == "" {
+			return nil, fmt.Errorf("hostPath or containerPath is empty in the CDI mount: %v", *mount)
+		}
+
+		chosenOpts = []string{}
+		for _, opt := range mount.Options {
+			if !shared.ValueInSlice(opt, chosenOpts) {
+				chosenOpts = append(chosenOpts, opt)
+			}
+		}
+
+		chosenOptsStr := strings.Join(chosenOpts, ",")
+		splittedDevName = strings.Split(mount.HostPath, "/")
+		deviceName = splittedDevName[len(splittedDevName)-1]
+
+		// mount.HostPath can be a symbolic link, so we need to evaluate it
+		evaluatedHostPath, err := filepath.EvalSymlinks(mount.HostPath)
+		if err != nil {
+			return nil, err
+		}
+
+		if evaluatedHostPath != mount.HostPath && mount.ContainerPath == strings.TrimPrefix(mount.HostPath, rootPath) {
+			indirectSymlinks = append(indirectSymlinks, cdi.SymlinkEntry{Target: strings.TrimPrefix(evaluatedHostPath, rootPath), Link: mount.ContainerPath})
+			mount.ContainerPath = strings.TrimPrefix(evaluatedHostPath, rootPath)
+		}
+
+		nativeDevices[deviceName] = map[string]string{
+			"type":              "disk",
+			"source":            evaluatedHostPath,
+			"path":              mount.ContainerPath,
+			"shift":             shift,
+			"raw.mount.options": chosenOptsStr,
+		}
+	}
+
+	// If the user desires to run a nested docker container inside a LXD container,
+	// the Tegra CSV files also need to be mounted so that the nvidia docker runtime
+	// can be auto-enabled as 'csv' mode.
+	if cdiID.Vendor() == cdi.Nvidia && cdiID.DeviceType() == cdi.IGPU {
```

Comment:
Question: If we construct a CDI spec that includes these files, then there should be no need to handle these separately.